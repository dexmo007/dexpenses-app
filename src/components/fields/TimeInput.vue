<template>
  <!-- <input
    type="text"
    v-mask="'##:##:##'"
    :value="value"
    @input="$emit('input', $event.target.value)"
  > -->
  <v-text-field
    :value="formattedValue"
    @input="emit"
    label="Time"
    :prepend-icon="icon"
    v-mask="'##:##:##'"
    :rules="[time]"
  ></v-text-field>
</template>
<script>
import { mask } from 'vue-the-mask';
import { fields } from '@/util/receipt';
import { time } from '@/util/form-rules';

export default {
  name: 'TimeInput',
  props: ['value'],
  directives: { mask },
  data() {
    return {
      time,
      icon: fields.time.icon,
      modelValue: null,
    };
  },
  computed: {
    formattedValue() {
      return this.format(this.value);
    },
  },
  methods: {
    emit(value) {
      if (!this.validate(value)) {
        return;
      }

      this.$emit('input', this.parse(value));
    },
    validate(s) {
      if (!s) {
        return true;
      }
      return !!s.match(/^([01]\d|2[0-4]):([0-5]\d):?(([0-5]\d))?$/);
    },
    format(d) {
      return this.$options.filters.time(d);
    },
    parse(s) {
      if (!s) {
        return null;
      }
      const [hour, minute, second] = s
        .split(':')
        .map(v => (v ? parseInt(v, 10) : null));
      return { hour, minute, second };
    },
  },
};
</script>
