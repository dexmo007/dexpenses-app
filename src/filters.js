import Vue from 'vue';

const dataSizes = ['B', 'KB', 'MB', 'GB', 'TB'];

function pickUnit(value) {
  let i = 0;
  let unitValue = value;
  while (unitValue > 1000 && i < dataSizes.length) {
    unitValue /= 1000;
    i += 1;
  }
  return [i, unitValue];
}

Vue.filter('dataSize', (value) => {
  const [i, unitValue] = pickUnit(value);
  return `${unitValue.toFixed(2)} ${dataSizes[i]}`;
});

Vue.filter('dataSizeProgress', ([value, total]) => {
  const [i, totalUnitValue] = pickUnit(total);
  return `${(value / (1000 ** i)).toFixed(2)}/${totalUnitValue.toFixed(2)} ${dataSizes[i]}`;
});

Vue.filter('humanReadableTime', (value) => {
  if (!value) {
    return '';
  }
  let date = value;
  if (value.toDate) {
    date = value.toDate();
  }
  return date.toLocaleDateString();
});

const currencies = {
  EUR: 'â‚¬',
};
Vue.filter('currency', (money) => {
  if (!money) {
    return '';
  }
  const { value, currency } = money;
  return `${value.toFixed(2)} ${currencies[currency] || currency}`;
});
