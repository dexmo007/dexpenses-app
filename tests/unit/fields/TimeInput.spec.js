import { mount, createLocalVue } from '@vue/test-utils';
import createVuetify from '../vuetify';
import TimeInput from '@/components/fields/TimeInput.vue';

import '@/filters';

const localVue = createLocalVue();

describe('TimeInput.vue', () => {
  let vuetify;
  beforeEach(() => {
    vuetify = createVuetify();
  });

  it('should render correctly', () => {
    const wrapper = mount(TimeInput, { localVue, vuetify });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render correctly with value', () => {
    const wrapper = mount(TimeInput, {
      propsData: {
        value: {
          hour: 12,
          minute: 0,
          second: 30,
        },
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('input').element.value).toEqual('12:00:30');
  });

  // it('should provide a masked input', async () => {
  //   const wrapper = mount(
  //     {
  //       template: '<TimeInput v-model="data"/>',
  //       data: () => ({ data: null }),
  //       components: { TimeInput },
  //     },
  //     { localVue, vuetify }
  //   );

  //   const input = wrapper.find('input');
  //   input.setValue('1200');
  //   input.trigger('keydown');
  //   console.log(wrapper.vm.$data.data);

  //   expect(input.element.value).toBe('12:0');
  // });

  it('should trigger input event if user selects new item', async () => {
    const wrapper = mount(TimeInput, { localVue, vuetify });
    wrapper.find('input').setValue('13:00:00');
    expect(wrapper.find('input').element.value).toEqual('13:00:00');
    await wrapper.vm.$nextTick();
    // nextTick handling is necessary bcs of mask attr on text-field (damy)
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted().input).toBeTruthy();
    expect(wrapper.emitted().input[0]).toHaveLength(1);
    expect(wrapper.emitted().input[0][0]).toEqual({ hour: 13, minute: 0, second: 0 });
  });

  it('should not trigger input event if new value is set programmatically', () => {
    const wrapper = mount(TimeInput, {
      propsData: {
        value: {
          hour: 12,
          minute: 30,
          second: 15,
        },
      },
      localVue,
      vuetify,
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.find('input').element.value).toEqual('12:30:15');
    wrapper.setProps({
      value: {
        hour: 13,
        minute: 30,
        second: 15,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.emitted().input).toBeFalsy();
    expect(wrapper.find('input').element.value).toEqual('13:30:15');
  });

  it('should update v-model correctly', async () => {
    const wrapper = mount(
      {
        template: '<TimeInput v-model="modelValue" />',
        data: () => ({
          modelValue: null,
        }),
        components: {
          TimeInput,
        },
      },
      { localVue, vuetify }
    );

    wrapper.find('input').setValue('13:00:00');
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.modelValue).toEqual({ hour: 13, minute: 0, second: 0 });
    expect(wrapper.element).toMatchSnapshot();
  });
});
