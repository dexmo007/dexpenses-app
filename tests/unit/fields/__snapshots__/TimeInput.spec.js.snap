// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput.vue should not trigger input event if new value is set programmatically 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label v-label--active theme--light"
          for="input-24"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-24"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeInput.vue should not trigger input event if new value is set programmatically 2`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label v-label--active theme--light"
          for="input-24"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-24"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeInput.vue should render correctly 1`] = `
<div
  class="v-input theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label theme--light"
          for="input-3"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-3"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeInput.vue should render correctly with value 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label v-label--active theme--light"
          for="input-10"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-10"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeInput.vue should trigger input event if user selects new item 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label v-label--active theme--light"
          for="input-17"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-17"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimeInput.vue should update v-model correctly 1`] = `
<div
  class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field"
>
  <div
    class="v-input__prepend-outer"
  >
    <div
      class="v-input__icon v-input__icon--prepend"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate material-icons theme--light"
      >
        access_time
      </i>
    </div>
  </div>
  <div
    class="v-input__control"
  >
    <div
      class="v-input__slot"
    >
      <div
        class="v-text-field__slot"
      >
        <label
          class="v-label v-label--active theme--light"
          for="input-32"
          style="left: 0px; position: absolute;"
        >
          Time
        </label>
        <input
          id="input-32"
          type="text"
        />
      </div>
    </div>
    <div
      class="v-text-field__details"
    >
      <div
        class="v-messages theme--light"
      >
        <span
          class="v-messages__wrapper"
          name="message-transition"
          tag="div"
        />
      </div>
    </div>
  </div>
</div>
`;
